[webkitapipy] Rework input files to recognize textual SPI and reduce configuration

emw-apple opened the pull request:
#### 5f96054d1d1a4ad40ccea8dd353b2e7c2b6ddb01
<pre>
[webkitapipy] Rework input files to recognize textual SPI and reduce configuration
<a href="https://bugs.webkit.org/show_bug.cgi?id=294760">https://bugs.webkit.org/show_bug.cgi?id=294760</a>
<a href="https://rdar.apple.com/153918728">rdar://153918728</a>

Reviewed by Brianna Fan.

A grab bag of changes to audit-spi&apos;s command line:

- Instead of being invoked with one &quot;primary file&quot; to audit and other
  files to allow IPI (arbitrary) usage of, support multiple input files,
  and check them in one invocation. This works towards being able to
  exhaustively check an allowlist to ensure that allowed SPI is cleaned
  up once it is no longer used.

- To indicate libraries for allowed IPI use, accept ld-style
  `-framework` and `-lx` argument to load frameworks and libraries on
  top of the base SDKDB, for allowed IPI use.

  For example, WebKit is allowed to use any symbol or declaration from
  JavaScriptCore and WebCore, so it is audited with `-framework WebCore
  -framework JavaScriptCore`.

- Some ObjC interfaces are headers-only. Namely, they are part of a
  framework but do not have any binary representation in that framework,
  because they are intended for clients to implement. For example,
  WebKit has many delegate protocols that it never adopts locally, but
  will call methods on. These selectors are flagged as unknown SPI.

  Fix by reordering when audit-spi runs, so that it runs *after* a
  target builds and its InstallAPI postprocessing tasks have run.
  Add a flag to OTHER_TAPI_FLAGS which instructs tapi to dump a partial
  SDKDB of the framework&apos;s headers. Teach audit-spi to find these
  partial SDKDBs and ingest them.

  For example: WebKit now produces a `WebKit.partial.sdkdb` containing
  records of its headers-only declarations. After WebKit, WebKitSwift,
  and _WebKit_SwiftUI finish building, audit-spi runs and checks all
  three. It finds the partial SDKDB and adds it to the internal
  database.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Emit a
  partial SDKDB during InstallAPI.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
  Update audit-spi invocation and move it to the aggregate target that
  builds after JSC has finished.
* Source/WTF/Scripts/audit-spi-if-needed.sh: Update audit-spi
  invocation.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Update audit-spi
  invocation.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj: Update audit-spi
  invocation.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Update audit-spi
  invocation, and move it to the aggregate target than runs after
  WebKit, WebKitSwift, and _WebKit_SwiftUI have finished building.
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj: Update
  audit-spi invocation.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
(Options): Added. Since there are a bunch of changes to the CLI, add a
helper object which provides type hints for all the arguments we
recognize. This makes future changes to the parser participate in
type checking.
(TSVReporter): Print file names in output when more than there&apos;s more
than one input file.
(get_parser): Add new arguments.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program_unittest.py: Added.

Canonical link: <a href="https://commits.webkit.org/296823@main">https://commits.webkit.org/296823@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f6546148c40fe2a9d776c1cb0ad88f40518d5b3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109712 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29370 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19799 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115733 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/59946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111675 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30048 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37958 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/83376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/59946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112660 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23948 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/98808 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63836 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/109144 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/23327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/16953 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59527 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/102203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/93323 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16990 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118525 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/108264 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36751 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/27220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92394 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37124 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95067 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/92216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23487 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37167 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/14909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36645 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42116 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132540 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36306 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/35887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39648 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/38015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
```diff
@@ -135,14 +151,32 @@ def finished(self):
+
+    binaries = parser.add_argument_group('framework and library dependencies',
+                                         description='''ld-style arguments to
+                                         support finding and allowing use of
```
super nit: Could this be "support finding and using binaries.." or "support finding and allow use of..."? "Support allowing use of" is a bit odd and hard to parse

briannafan commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
```diff
@@ -135,14 +151,32 @@ def finished(self):
+                                         local build products.''')
+
+    binaries.add_argument('-framework', metavar='FRAMEWORK', type=str,
+                          action='append', dest='frameworks',
```
nit: For readability, it would be nice to have all the parameters in the same order/line. `-framework` and `-l` orders are similar but getting to `-F` and `-L` broke my brain a little bit

briannafan reviewed (approved):
Not much other than two nits. Appreciate your help explaining these changes in the context of the project as a whole :)

(Reviewing this has convinced me to add type hints in my future work!!)

emw-apple committed:
[webkitapipy] Rework input files to recognize textual SPI and reduce configuration
https://bugs.webkit.org/show_bug.cgi?id=294760
rdar://153918728

Reviewed by Brianna Fan.

A grab bag of changes to audit-spi's command line:

- Instead of being invoked with one "primary file" to audit and other
  files to allow IPI (arbitrary) usage of, support multiple input files,
  and check them in one invocation. This works towards being able to
  exhaustively check an allowlist to ensure that allowed SPI is cleaned
  up once it is no longer used.

- To indicate libraries for allowed IPI use, accept ld-style
  `-framework` and `-lx` argument to load frameworks and libraries on
  top of the base SDKDB, for allowed IPI use.

  For example, WebKit is allowed to use any symbol or declaration from
  JavaScriptCore and WebCore, so it is audited with `-framework WebCore
  -framework JavaScriptCore`.

- Some ObjC interfaces are headers-only. Namely, they are part of a
  framework but do not have any binary representation in that framework,
  because they are intended for clients to implement. For example,
  WebKit has many delegate protocols that it never adopts locally, but
  will call methods on. These selectors are flagged as unknown SPI.

  Fix by reordering when audit-spi runs, so that it runs *after* a
  target builds and its InstallAPI postprocessing tasks have run.
  Add a flag to OTHER_TAPI_FLAGS which instructs tapi to dump a partial
  SDKDB of the framework's headers. Teach audit-spi to find these
  partial SDKDBs and ingest them.

  For example: WebKit now produces a `WebKit.partial.sdkdb` containing
  records of its headers-only declarations. After WebKit, WebKitSwift,
  and _WebKit_SwiftUI finish building, audit-spi runs and checks all
  three. It finds the partial SDKDB and adds it to the internal
  database.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Emit a
  partial SDKDB during InstallAPI.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
  Update audit-spi invocation and move it to the aggregate target that
  builds after JSC has finished.
* Source/WTF/Scripts/audit-spi-if-needed.sh: Update audit-spi
  invocation.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Update audit-spi
  invocation.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj: Update audit-spi
  invocation.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Update audit-spi
  invocation, and move it to the aggregate target than runs after
  WebKit, WebKitSwift, and _WebKit_SwiftUI have finished building.
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj: Update
  audit-spi invocation.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
(Options): Added. Since there are a bunch of changes to the CLI, add a
helper object which provides type hints for all the arguments we
recognize. This makes future changes to the parser participate in
type checking.
(TSVReporter): Print file names in output when more than there's more
than one input file.
(get_parser): Add new arguments.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program_unittest.py: Added.

Canonical link: https://commits.webkit.org/296823@main
